crates/ra_assists/src/assists/apply_demorgan.rs

```diff
@@ -0,0 +1,102 @@
+//! This contains the functions associated with the demorgan assist.
+//! This assist transforms boolean expressions of the form `!a || !b` into
+//! `!(a && b)`.
+use hir::db::HirDatabase;
+use ra_syntax::ast::{AstNode, BinExpr, BinOp, Expr, PrefixOp};
+use ra_syntax::SyntaxNode;
+
+use crate::{Assist, AssistCtx, AssistId};
+
+// Return the opposite text for a given logical operator, if it makes sense
+fn opposite_logic_op(kind: BinOp) -> Option<&'static str> {
+    match kind {
+        BinOp::BooleanOr => Some("&&"),
+        BinOp::BooleanAnd => Some("||"),
+        _ => None,
+    }
+}
+
+// This function tries to undo unary negation, or inequality
+fn undo_negation(node: SyntaxNode) -> Option<String> {
+    match Expr::cast(node)? {
+        Expr::BinExpr(bin) => match bin.op_kind()? {
+            BinOp::NegatedEqualityTest => {
+                let lhs = bin.lhs()?.syntax().text();
+                let rhs = bin.rhs()?.syntax().text();
+                Some(format!("{} == {}", lhs, rhs))
+            }
+            _ => None,
+        },
+        Expr::PrefixExpr(pe) => match pe.op_kind()? {
+            PrefixOp::Not => {
+                let child = pe.expr()?.syntax().text();
+                Some(String::from(child))
+            }
+            _ => None,
+        },
+        _ => None,
+    }
+}
+
+/// Assist for applying demorgan's law
+///
+/// This transforms expressions of the form `!l || !r` into `!(l && r)`.
+/// This also works with `&&`. This assist can only be applied with the cursor
+/// on either `||` or `&&`, with both operands being a negation of some kind.
+/// This means something of the form `!x` or `x != y`.
```

Comment:
Yeah, I think this is a good start, but, long term, we should lean to negate arbitrary expressions, to transform `x || !y`  into `!(!x && y)`

Reply:
One of the main tricky things I ran into when working on this aspect is figuring out when you need `()` around the expression you're putting a `!` on. I also couldn't figure out how to match against a litteral `true` or `false` respectively.

Reply:
Yeah, to implement this fully quite a few special cases need to be handled. IntelliJ impl is here, although I believe it is also incomplete: 

https://github.com/intellij-rust/intellij-rust/blob/659f7410937ccc382ec324af1c8c647a2e233e50/src/main/kotlin/org/rust/ide/intentions/DemorgansLawIntention.kt